### ISSUE TITLE
Exception when saving on folderish type, hard to track

### ISSUE THREAD

**Issue by thet** [plone/Products.CMFEditions#26]
i got this exception on a folderish type document (folderish type, default page of a parent folder, has 4 images within):

```
Traceback (innermost last):

    Module ZPublisher.Publish, line 138, in publish
    Module ZPublisher.mapply, line 77, in mapply
    Module ZPublisher.Publish, line 48, in call_object
    Module plone.z3cform.layout, line 66, in __call__
    Module plone.z3cform.layout, line 50, in update
    Module plone.dexterity.browser.edit, line 62, in update
    Module plone.z3cform.fieldsets.extensible, line 59, in update
    Module plone.z3cform.patch, line 30, in GroupForm_update
    Module z3c.form.group, line 145, in update
    Module plone.app.z3cform.csrf, line 21, in execute
    Module z3c.form.action, line 98, in execute
    Module z3c.form.button, line 315, in __call__
    Module z3c.form.button, line 170, in __call__
    Module plone.dexterity.browser.edit, line 26, in handleApply
    Module z3c.form.group, line 126, in applyChanges
    Module zope.event, line 31, in notify
    Module zope.component.event, line 24, in dispatch
    Module zope.component._api, line 136, in subscribers
    Module zope.component.registry, line 321, in subscribers
    Module zope.interface.adapter, line 585, in subscribers
    Module zope.component.event, line 32, in objectEventNotify
    Module zope.component._api, line 136, in subscribers
    Module zope.component.registry, line 321, in subscribers
    Module zope.interface.adapter, line 585, in subscribers
    Module plone.app.versioningbehavior.subscribers, line 62, in create_version_on_save
    Module Products.CMFEditions.CopyModifyMergeRepositoryTool, line 305, in save
    Module Products.CMFEditions.CopyModifyMergeRepositoryTool, line 451, in _recursiveSave
    Module Products.CMFEditions.ArchivistTool, line 267, in prepare
    Module Products.CMFEditions.ArchivistTool, line 221, in _cloneByPickle

TypeError: Can't pickle objects in acquisition wrappers.
```

I placed a pdb in a try/except block here: `Module Products.CMFEditions.ArchivistTool, line 221, in _cloneByPickle`. It fails at this line `p.dump(aq_base(obj))`. So, here is a aq_base statement, which should return the object with all acquisition wrappers removed. When I run into this problem with pdb, I tried the same statement and it worked. Weired. I suspect it has something to do with the folderish structure, but on the other hand this is the first time I encounter this problem, even with many collective.folderishtypes deployments since years.
However, I'm going to deactivate versioningbehavior for a quick fix.

**Comment by vincentfretin** [comment]
@cedricmessiant, I think we had this exact traceback, the reason was a field that contains RelationValue. How did we fix it, do you remember?
@thet Do you have a RelationChoice on your content type by any chance?

**Comment by Gagaro** [comment]
I have the same issue with a content type with a RelationChoice.

Do you remember how you fixed this ?

**Comment by cedricmessiant** [comment]
When I look at the commit that seems related to the fix in our product, it seems that I just : 

- added a `repositorytool.xml` file in our profile
- used `Compound Diff for Dexterity types` for my content type: `api.portal.get_tool('portal_diff').setDiffForPortalType( 'mycontenttype', {'any': "Compound Diff for Dexterity types"})`

**Comment by Gagaro** [comment]
Thanks a lot, it works for me as well :+1: .

**Comment by cedricmessiant** [comment]
Cool !

**Comment by esoth** [comment]
I'm having the same issue and updating repositorytool.xml and diff_too.xml profiles (and running steps) does not help. In fact, I already had those steps installed for my dexterity content type. It doesn't make sense that this would matter here anyway. Do you know what version of dexterity and Plone you were using?

**Comment by Gagaro** [comment]
I was on either Plone 5.0b2 or 5.0. It was the default dexterity version.

2015-12-23 18:10 GMT+01:00 esoth <notifications@github.com>:

> I'm having the same issue and updating repositorytool.xml and diff_too.xml
> profiles (and running steps) does not help. In fact, I already had those
> steps installed for my dexterity content type. It doesn't make sense that
> this would matter here anyway. Do you know what version of dexterity and
> Plone you were using?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/plone/Products.CMFEditions/issues/26#issuecomment-166946826>
> .
>



-- 
Gagaro

**Comment by darryldixon** [comment]
I have tracked this issue down to the fact that for some reason, CMFEditions is trying to pickle the OFS children of the folderish item. The error thrown "TypeError: Can't pickle objects in acquisition wrappers." is being emitted from C code in _Acquisition.c. I instrumented this code so that it would include a __repr__ of the object it was complaining about. This showed that it was the sub-objects (OFS children) of the item being versioned that the error was being emitted for.

In other words, for some reason the OMOutsideChildrensModifier is not preventing child items from being included in the pickler stream. This is a CMFEditions/Dexterity bug, one way or the other.

**Comment by darryldixon** [comment]
For those that want to see this for themselves, you will need to manually patch the Acquisition egg and then insert in to your buildout, eg:

`diff -ru Acquisition-4.7/setup.py Acquisition-4.7a/setup.py
--- Acquisition-4.7/setup.py	2020-10-07 22:44:28.000000000 +1300
+++ Acquisition-4.7a/setup.py	2022-04-06 18:29:55.235552924 +1200
@@ -36,7 +36,7 @@
                   include_dirs=['include', 'src']),
     ]
 
-version = '4.7'
+version = '4.7a'
 
 setup(
     name='Acquisition',
diff -ru Acquisition-4.7/src/Acquisition/_Acquisition.c Acquisition-4.7a/src/Acquisition/_Acquisition.c
--- Acquisition-4.7/src/Acquisition/_Acquisition.c	2020-10-07 22:44:28.000000000 +1300
+++ Acquisition-4.7a/src/Acquisition/_Acquisition.c	2022-04-07 00:14:47.942013701 +1200
@@ -1477,8 +1477,16 @@
 PyObject *
 Wrappers_are_not_picklable(PyObject *wrapper, PyObject *args)
 {
+    PyObject *obj;
+    PyObject *repr;
+    /* C progammers disease... */
+    char msg[1024];
+    /* Unwrap wrapper completely -> obj. */
+    obj = get_base(wrapper);
+    repr = PyObject_Repr((PyObject *)obj);
+    snprintf(msg, 1024, "Can't pickle objects in acquisition wrappers - %s", (char *)PyBytes_AS_STRING(repr));
     PyErr_SetString(PyExc_TypeError,
-                    "Can't pickle objects in acquisition wrappers.");
+                    msg);
     return NULL;
 }
 
`

**Comment by darryldixon** [comment]
I have solved what is causing this issue but do not yet have a good solution:

Dexterity items that use z3c.relationfield.schema RelationList or RelationChoice, (eg, they store a RelationValue) will trigger this error. This is because RelationValue items use an attribute of __parent__ which points back to the original Dexterity item. When CMFEditions tries to clone the Dexterity item with the Python Pickler, the Pickler follows the __parent__ attribute back around to the original item, except this time wrapped in an Acquisition wrapper, which then fails with the error recorded above. Presumably the Pickler normally has mechanisms to avoid circular references; but because following the __parent__ results in the Acquisition wrapper, the Pickler attempts to pickle the wrapper, which throws TypeError.

**Comment by davisagli** [comment]
RelationValues should already be skipped by the SkipRelations modifier which is implemented in plone.app.versioningbehavior: https://github.com/plone/plone.app.versioningbehavior/blob/08637c3e518fdbf38a158795d9a3bfad1abf5e36/plone/app/versioningbehavior/modifiers.py#L170

I think I found the bug in OMOutsideChildrensModifier which makes it fail to skip child items if they are acquisition-wrapped. PR coming...

**Comment by darryldixon** [comment]
> RelationValues should already be skipped by the SkipRelations modifier which is implemented in plone.app.versioningbehavior: https://github.com/plone/plone.app.versioningbehavior/blob/08637c3e518fdbf38a158795d9a3bfad1abf5e36/plone/app/versioningbehavior/modifiers.py#L170
> 
> I think I found the bug in OMOutsideChildrensModifier which makes it fail to skip child items if they are acquisition-wrapped. PR coming...

This is probably true, but Dexterity allows for easy creation of complex types that can end up with RelationValues etc embedded. Probably needs a note in the developer docs somewhere to the effect that anyone doing this will need to do careful analysis and possibly the development of a custom modifier to avoid problems. Ideally, the `__parent__` attribute on RelationValue would be simply removed. Basically, if you have any Type anywhere that embeds any other Type, you run the risk of exploding with this error when you try to Version the top-level item.

**Comment by davisagli** [comment]
@darryldixon Yeah, it would be nice if we could find a way to handle these regardless of where they are located. It's not quite so simple as just removing the RelationValues and/or parent pointers when the object is archived, because they also need to be restored when an archived revision is restored -- which means the modifier needs to know where they are.
